core/domain/prod_validation_jobs_one_off.py

```diff
@@ -2474,213 +2600,1326 @@ def _get_custom_validation_functions(cls):
-    skill_models.SkillRightsSnapshotContentModel: (
-        SkillRightsSnapshotContentModelValidator),
-    skill_models.SkillCommitLogEntryModel: (
-        SkillCommitLogEntryModelValidator),
-    skill_models.SkillSummaryModel: SkillSummaryModelValidator,
-    story_models.StoryModel: StoryModelValidator,
-    story_models.StorySnapshotMetadataModel: (
-        StorySnapshotMetadataModelValidator),
-    story_models.StorySnapshotContentModel: (
-        StorySnapshotContentModelValidator),
-    story_models.StoryRightsModel: StoryRightsModelValidator,
-    story_models.StoryRightsSnapshotMetadataModel: (
-        StoryRightsSnapshotMetadataModelValidator),
-    story_models.StoryRightsSnapshotContentModel: (
-        StoryRightsSnapshotContentModelValidator),
-    story_models.StoryCommitLogEntryModel: (
-        StoryCommitLogEntryModelValidator),
-    story_models.StorySummaryModel: StorySummaryModelValidator,
-    user_models.UserSubscriptionsModel: UserSubscriptionsModelValidator,
-}
-
+        Args:
+            item: ndb.Model. UserSettingsModel to validate.
+        """
+        time_fields = {
+            'last agreed to terms': item.last_agreed_to_terms,
+            'last started state editor tutorial': (
+                item.last_started_state_editor_tutorial),
+            'last started state translation tutorial': (
+                item.last_started_state_translation_tutorial),
+            'last logged in': item.last_logged_in,
+            'last edited an exploration': item.last_edited_an_exploration,
+            'last created an exploration': item.last_created_an_exploration
+        }
+        current_time = datetime.datetime.utcnow()
+        for time_field_name, time_field_value in time_fields.iteritems():
+            if time_field_value and time_field_value > current_time:
+                cls.errors['%s check' % time_field_name].append(
+                    'Entity id %s: Value for %s: %s is greater than the '
+                    'time when job was run' % (
+                        item.id, time_field_name, time_field_value))
+
+        current_msec = utils.get_current_time_in_millisecs()
+        if item.first_contribution_msec and (
+                item.first_contribution_msec > current_msec):
+            cls.errors['first contribution check'].append(
+                'Entity id %s: Value for first contribution msec: %s is '
+                'greater than the time when job was run' % (
+                    item.id, item.first_contribution_msec))
+
+    @classmethod
+    def _validate_normalized_username_is_unique(cls, item):
+        """Validates that normalized username is unique.
 
-class ProdValidationAuditOneOffJob(jobs.BaseMapReduceOneOffJobManager):
-    """Job that audits and validates production models."""
+        Args:
+            item: ndb.Model. UserSettingsModel to validate.
+        """
+        user_models_list = user_models.UserSettingsModel.query().filter(
```

Comment:
This seems like a bad idea -- you're doing one query per user model, when you could instead just do one projection query for all the models, extract all the normalized usernames, sort them, and figure out if there are any duplicates.

Reply:
So, should this be done in a separate job instead since it will not bendependent on a single model instance now?

Reply:
Probably yes. I'm not sure if this is the case for others but it might be worth checking to see if you have similar patterns elsewhere. Doing N queries is quite resource-intensive.

Reply:
I was trying to get all user models in a global list and then use it as follows:

```
ALL_USER_MODELS = user_models.UserSettingsModel.get_all().fetch()
USER_IDS_AND_NORMALIZED_USERNAMES = [
        (user_model.id, user_model.normalized_username)
        for user_model in ALL_USER_MODELS if user_model is not None]
```

But when I run the test, I get the error: `WARNING:root:initial generator _run_to_list(query.py:987) raised BadArgumentError(app must not be empty.)`. Is there any other way to fetch all the user models once for the job? 

I also found a solution for testing: https://www.dangtrinh.com/2013/12/google-app-engine-sdk-for-python-using.html but I am not sure if we should do this.

Reply:
The solution mentioned in the last paragraph seems hacky, let's not do it.

For the first part -- I don't understand where you're adding the code. Is it at module-level? If so, that doesn't seem so good, it will get run when the module is imported which is probably not what you want.

Perhaps your best bet is to just have a special MR job that iterates through all the models and that yields the normalized usernames. Then have reduce() check that there are no duplicates.

Reply:
Yes, I was adding the code at module level but I have changed it now to the solution you mentioned. Thanks for the solution!